human-readable-types/src/main/java/com/palantir/humanreadabletypes/StringBasedUrl.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.humanreadabletypes;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+/**
+ * An immutable {@link URL} whose {@link #equals}, {@link #toString}, and {@link #hashCode} are derived from the URL's
+ * string representation (at instance construction time) rather than the {@link URL} class. For example, two {@link
+ * StringBasedUrl}s are equal iff the strings from which they were {@link #valueOf constructed} are equal.
+ * <p>
+ * This class is intended to be used in configuration files where (1) URL syntax verification should be performed, (2)
+ * host name resolution is not desired, and (3) equality/stability of the configuration file should be derived from the
+ * configuration file syntax, rather than the dynamic behavior of the configuration values, cf. {@link URL#equals}.
+ */
+public final class StringBasedUrl {
+
+    private final String stringUrl;
+    private final URL url;
+
+    private StringBasedUrl(String stringUrl) throws MalformedURLException {
+        this.stringUrl = stringUrl;
+        this.url = new URL(stringUrl);
+    }
+
+    /**
+     * Constructs a new {@link StringBasedUrl} from the provided string representation, or throws an {@link
+     * IllegalArgumentException} if the given string cannot be parsed as a {@link URL}.
+     */
+    @JsonCreator
+    public static StringBasedUrl valueOf(String url) {
+        try {
+            return new StringBasedUrl(url);
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("Illegal URL", e);
```

Comment:
for config-friendliness I think it might be nicer to defer the exception throwing to the `toUrl` method, which should make it less likely that a bad URL causes an entire configuration not to parse, and instead localizes the problem wherever is consuming the URL.